src/System.Binary.Base64/System/Binary/Base64.cs

```diff
@@ -149,82 +191,160 @@ public static void Decode(byte b0, byte b1, byte b2, byte b3, out byte r0, out b
             byte i2 = s_decodingMap[b2];
             byte i3 = s_decodingMap[b3];
 
-            r0 = (byte)(i0<<2 | i1 >> 4);
-            r1 = (byte)(i1<<4 | i2>>2);
-            r2 = (byte)(i2<<6 | i3);
+            r0 = (byte)(i0 << 2 | i1 >> 4);
+            r1 = (byte)(i1 << 4 | i2 >> 2);
+            r2 = (byte)(i2 << 6 | i3);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int Decode(ref byte srcBytes)
+        {
+            byte b0 = srcBytes;
+            byte b1 = Unsafe.Add(ref srcBytes, 1);
+            byte b2 = Unsafe.Add(ref srcBytes, 2);
+            byte b3 = Unsafe.Add(ref srcBytes, 3);
+
+            Decode(b0, b1, b2, b3, out byte r0, out byte r1, out byte r2);
+
+            int result = r2 << 16 | r1 << 8 | r0;
+            return result;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Decode(byte b0, byte b1, byte b2, out byte r0, out byte r1)
+        {
+            byte i0 = s_decodingMap[b0];
+            byte i1 = s_decodingMap[b1];
+            byte i2 = s_decodingMap[b2];
+
+            r0 = (byte)(i0 << 2 | i1 >> 4);
+            r1 = (byte)(i1 << 4 | i2 >> 2);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int DecodePadByOne(ref byte srcBytes)
+        {
+            byte b0 = srcBytes;
+            byte b1 = Unsafe.Add(ref srcBytes, 1);
+            byte b2 = Unsafe.Add(ref srcBytes, 2);
+
+            Decode(b0, b1, b2, out byte r0, out byte r1);
+
+            int result = r1 << 8 | r0;
+            return result;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void Decode(byte b0, byte b1, out byte r0)
+        {
+            byte i0 = s_decodingMap[b0];
+            byte i1 = s_decodingMap[b1];
+
+            r0 = (byte)(i0 << 2 | i1 >> 4);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int DecodePadByTwo(ref byte srcBytes)
+        {
+            byte b0 = srcBytes;
+            byte b1 = Unsafe.Add(ref srcBytes, 1);
+
+            Decode(b0, b1, out byte r0);
+
+            int result = r0;
+            return result;
         }
 
         /// <summary>
         /// 
         /// </summary>
-        /// <param name="buffer"></param>
-        /// <returns>Number of bytes written to the buffer.</returns>
-        public static int DecodeInPlace(Span<byte> buffer)
+        /// <param name="source"></param>
+        /// <param name="destination"></param>
+        /// <returns>Number of bytes written to the destination.</returns>
+        public static int Decode(ReadOnlySpan<byte> source, Span<byte> destination)
         {
-            int di = 0;
-            int si = 0;
-            byte r0, r1, r2;
+            //Diagnostics.Debug.Assert(destination.Length >= ComputeDecodedLength(source.Length));
```

Comment:
There is a destination buffer overflow here if destination isn't large enough to fit the decoded source.